File: gateway/middleware/ipallowlist.go
package middleware

// IP 白名单中间件 — 与应用层 JWT/网络层防火墙形成纵深防御:
//   - 默认放行本地回环 (127.0.0.0/8, ::1)
//   - 通过 IP_ALLOWLIST 环境变量声明跨主机授信网段 (逗号分隔, 支持 CIDR)
//   - TRUST_PROXY=true 时才信任 X-Forwarded-For, 否则只看 TCP 直连地址
//
// 强烈建议: 仅允许来自 127.0.0.1 + 显式配置的内网网段, 严禁公网 0.0.0.0/0.

import (
	"log"
	"net"
	"net/http"
	"os"
	"strings"

	"github.com/gin-gonic/gin"
)

var loopbackNets = mustParseCIDRs([]string{"127.0.0.0/8", "::1/128"})

// IPAllowlist 返回基于 CIDR 白名单的 Gin 中间件.
//
// allowlist: 允许网段 (本地回环始终放行)
// trustProxy: 是否优先采用 X-Forwarded-For (仅在前置可信反代时开启)
// exemptPaths: 健康检查等无需鉴权的路径
func IPAllowlist(allowlist []*net.IPNet, trustProxy bool, exemptPaths ...string) gin.HandlerFunc {
	nets := append([]*net.IPNet{}, allowlist...)
	nets = append(nets, loopbackNets...)
	exempt := make(map[string]struct{}, len(exemptPaths))
	for _, p := range exemptPaths {
		exempt[p] = struct{}{}
	}

	return func(c *gin.Context) {
		if _, ok := exempt[c.Request.URL.Path]; ok {
			c.Next()
			return
		}

		ip := clientIP(c, trustProxy)
		if ip == nil {
			c.AbortWithStatusJSON(http.StatusForbidden, gin.H{
				"code":    "FORBIDDEN",
				"message": "client ip unknown",
			})
			return
		}

		for _, n := range nets {
			if n.Contains(ip) {
				c.Next()
				return
			}
		}

		log.Printf("IPAllowlist: 拒绝来自 %s 的访问 -> %s", ip.String(), c.Request.URL.Path)
		c.AbortWithStatusJSON(http.StatusForbidden, gin.H{
			"code":    "FORBIDDEN",
			"message": "client ip not allowed",
		})
	}
}

// ParseAllowlist 把 "10.0.0.0/8,192.168.1.10" 解析为 *net.IPNet 列表.
// 单 IP 自动补全为 /32 或 /128.
func ParseAllowlist(raw string) []*net.IPNet {
	if raw == "" {
		return nil
	}
	out := make([]*net.IPNet, 0, 4)
	for _, item := range strings.Split(raw, ",") {
		item = strings.TrimSpace(item)
		if item == "" {
			continue
		}
		if !strings.Contains(item, "/") {
			if ip := net.ParseIP(item); ip != nil {
				if ip.To4() != nil {
					item += "/32"
				} else {
					item += "/128"
				}
			}
		}
		_, cidr, err := net.ParseCIDR(item)
		if err != nil {
			log.Printf("IPAllowlist: 跳过非法条目 %q: %v", item, err)
			continue
		}
		out = append(out, cidr)
	}
	return out
}

// AllowlistFromEnv 读取 IP_ALLOWLIST / TRUST_PROXY 环境变量.
func AllowlistFromEnv() ([]*net.IPNet, bool) {
	nets := ParseAllowlist(os.Getenv("IP_ALLOWLIST"))
	trust := strings.EqualFold(os.Getenv("TRUST_PROXY"), "true") ||
		os.Getenv("TRUST_PROXY") == "1"
	return nets, trust
}

func clientIP(c *gin.Context, trustProxy bool) net.IP {
	if trustProxy {
		if xff := c.GetHeader("X-Forwarded-For"); xff != "" {
			first := strings.TrimSpace(strings.Split(xff, ",")[0])
			if ip := net.ParseIP(first); ip != nil {
				return ip
			}
		}
		if real := c.GetHeader("X-Real-IP"); real != "" {
			if ip := net.ParseIP(strings.TrimSpace(real)); ip != nil {
				return ip
			}
		}
	}
	host, _, err := net.SplitHostPort(c.Request.RemoteAddr)
	if err != nil {
		host = c.Request.RemoteAddr
	}
	return net.ParseIP(host)
}

func mustParseCIDRs(items []string) []*net.IPNet {
	out := make([]*net.IPNet, 0, len(items))
	for _, item := range items {
		_, cidr, err := net.ParseCIDR(item)
		if err != nil {
			panic(err)
		}
		out = append(out, cidr)
	}
	return out
}

File: agent/middleware/ip_allowlist.py
"""IP 白名单中间件.

合规要求:
    - 服务端口默认仅监听 127.0.0.1, 严禁 0.0.0.0 全网监听.
    - 跨主机调用必须配置 "IP + 端口" 白名单, 仅放行授信地址.

本模块提供应用层的 IP 白名单, 与网络层 (iptables/安全组) 形成纵深防御.
支持单 IP / CIDR 网段配置, 并兼容 Docker / 反向代理场景下的 X-Forwarded-For.
"""
from __future__ import annotations

import ipaddress
import logging
import os
from typing import Iterable

from fastapi import Request
from fastapi.responses import JSONResponse
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.types import ASGIApp

logger = logging.getLogger(__name__)

# 始终放行的本地回环网段: IPv4 127.0.0.0/8 + IPv6 ::1
_LOOPBACK_NETS: tuple[ipaddress._BaseNetwork, ...] = (
    ipaddress.ip_network("127.0.0.0/8"),
    ipaddress.ip_network("::1/128"),
)


def _parse_networks(raw: str | None) -> list[ipaddress._BaseNetwork]:
    """解析逗号分隔的 IP / CIDR 列表为网段对象."""
    if not raw:
        return []
    nets: list[ipaddress._BaseNetwork] = []
    for item in raw.split(","):
        item = item.strip()
        if not item:
            continue
        try:
            nets.append(ipaddress.ip_network(item, strict=False))
        except ValueError as exc:
            logger.warning("IP_ALLOWLIST 跳过非法条目 %r: %s", item, exc)
    return nets


def _client_ip(request: Request, trust_proxy: bool) -> str | None:
    """提取客户端真实 IP.

    - trust_proxy=True 时优先取 X-Forwarded-For 链路首跳 (网关/反代场景);
    - 否则使用 TCP 直连地址, 防止伪造头绕过白名单.
    """
    if trust_proxy:
        xff = request.headers.get("x-forwarded-for")
        if xff:
            return xff.split(",")[0].strip()
        real = request.headers.get("x-real-ip")
        if real:
            return real.strip()
    if request.client is not None:
        return request.client.host
    return None


class IPAllowlistMiddleware(BaseHTTPMiddleware):
    """基于 IP / CIDR 的白名单中间件.

    Args:
        app: ASGI 应用.
        allowlist: 允许网段列表 (本地回环始终放行).
        trust_proxy: 是否信任 X-Forwarded-For (仅在前置可信反代时开启).
        exempt_paths: 健康检查等无需鉴权的路径.
    """

    def __init__(
        self,
        app: ASGIApp,
        allowlist: Iterable[ipaddress._BaseNetwork] = (),
        trust_proxy: bool = False,
        exempt_paths: Iterable[str] = ("/health",),
    ) -> None:
        super().__init__(app)
        self._nets: tuple[ipaddress._BaseNetwork, ...] = tuple(allowlist) + _LOOPBACK_NETS
        self._trust_proxy = trust_proxy
        self._exempt = tuple(exempt_paths)

    async def dispatch(self, request: Request, call_next):
        if request.url.path in self._exempt:
            return await call_next(request)

        client = _client_ip(request, self._trust_proxy)
        if client is None:
            return JSONResponse(
                status_code=403,
                content={"code": "FORBIDDEN", "message": "client ip unknown"},
            )

        try:
            ip_obj = ipaddress.ip_address(client)
        except ValueError:
            logger.warning("IPAllowlist: 非法客户端 IP %r", client)
            return JSONResponse(
                status_code=403,
                content={"code": "FORBIDDEN", "message": "invalid client ip"},
            )

        if not any(ip_obj in net for net in self._nets):
            logger.warning("IPAllowlist: 拒绝来自 %s 的访问 -> %s", client, request.url.path)
            return JSONResponse(
                status_code=403,
                content={"code": "FORBIDDEN", "message": "client ip not allowed"},
            )

        return await call_next(request)


def from_env(env_key: str = "IP_ALLOWLIST") -> tuple[list[ipaddress._BaseNetwork], bool]:
    """从环境变量读取白名单与 trust_proxy 配置."""
    nets = _parse_networks(os.getenv(env_key))
    trust_proxy = os.getenv("TRUST_PROXY", "false").lower() in ("1", "true", "yes")
    return nets, trust_proxy
